File: src/util/KM.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.StreamTokenizer;

public class KM {
    public static int INF = 0x7fffffff;

    public Main(int n, int graph[][]) {
        this.n = n;
        this.graph = graph;
        this.Lx = new int[n];
        this.Ly = new int[n];
        this.matched = new int[n];
        for (int i = 0; i < n; i++) {
            this.Lx[i] = -INF;
            for (int j = 0; j < n; j++) {
                this.Lx[i] = Math.max(this.Lx[i], graph[i][j]);
            }
            this.Ly[i] = 0;
            this.matched[i] = -1;
        }
    }

    public int km() {
        for (int i = 0; i < n; i++) {
            slack = new int[n];
            for (int j = 0; j < n; j++)
                slack[j] = INF;
            while (true) {
                S = new boolean[n];
                T = new boolean[n];
                if (match(i))
                    break;
                else
                    update();
            }
        }
        int result = 0;
        for (int i = 0; i < n; i++) {
            if (matched[i] != -1)
                result += graph[matched[i]][i];
        }
        return result;
    }

    private boolean match(int u) {
        S[u] = true;
        for (int v = 0; v < n; v++) {
            if (T[v])
                continue;
            int t = Lx[u] + Ly[v] - graph[u][v];
            if (t == 0) {
                T[v] = true;
                if (matched[v] == -1 || match(matched[v])) {
                    matched[v] = u;
                    return true;
                }
            } else {
                slack[v] = Math.min(slack[v], t);
            }
        }
        return false;
    }

    private void update() {
        int d = INF;
        for (int i = 0; i < n; i++) {
            if (!T[i])
                d = Math.min(d, slack[i]);
        }
        for (int i = 0; i < n; i++)
            if (S[i])
                Lx[i] -= d;
        for (int i = 0; i < n; i++)
            if (T[i])
                Ly[i] += d;
            else
                slack[i] -= d;

    }

    private int n;
    private int graph[][];
    private int Lx[];
    private int Ly[];
    private int slack[];
    private boolean S[];
    private boolean T[];
    private int matched[];

    // http://acm.hdu.edu.cn/showproblem.php?pid=2255
    public static void main(String[] args) throws IOException {
        int maxn = 456;
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        int graph[][] = new int[maxn][maxn];
        String line;
        while (in.nextToken() != StreamTokenizer.TT_EOF) {
            int n = (int)in.nval;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    int a = in.nextToken();
                    graph[i][j] = (int) in.nval;
                }
            }

            System.out.println(new KM(n, graph).km());
        }
    }
}

File: src/util/KM.py
class KM:
    INF = 0x7fffffff

    def __init__(self, n, graph):
        self.n = n
        self.graph = graph  # adjacent matrix
        self.Lx = [max(self.graph[i]) for i in range(self.n)]
        self.Ly = [0] * n
        self.slack = None
        self.S = None
        self.T = None
        self.matched = [-1] * n

    def match(self, u):
        self.S[u] = True
        for v in range(self.n):
            # v is already in the cross road
            if self.T[v]:
                continue
            t = self.Lx[u] + self.Ly[v] - self.graph[u][v]
            if not t:
                self.T[v] = True
                if self.matched[v] == -1 or self.match(self.matched[v]):
                    self.matched[v] = u
                    return True
            else:
                self.slack[v] = min(self.slack[v], t)
        return False

    def update(self):
        d = min(self.slack[i] for i in range(self.n) if not self.T[i])
        for i in range(self.n):
            if self.S[i]:
                self.Lx[i] -= d
        for i in range(self.n):
            if self.T[i]:
                self.Ly[i] += d
            else:
                self.slack[i] -= d

    def km(self):
        for i in range(self.n):
            self.slack = [KM.INF] * self.n
            while True:
                self.S = [False] * self.n
                self.T = [False] * self.n
                if self.match(i):
                    break
                else:
                    self.update()
        return sum(self.graph[self.matched[i]][i] for i in range(self.n) if self.matched[i] != -1)


# http://codeforces.com/contest/321/problem/B
if __name__ == '__main__':
    inf = KM.INF
    maxn = 500
    graph = []
    for i in range(maxn):
        graph.append([0] * maxn)
    n, m = [int(x) for x in input().split()]
    jiro = [input().split() for i in range(n)]
    ciel = [int(input()) for i in range(m)]
    for i in range(m):
        for j in range(n):
            if jiro[j][0] == 'ATK':
                if ciel[i] >= int(jiro[j][1]):
                    graph[i][j] = ciel[i] - int(jiro[j][1])
    ans = KM(m, graph).km()

    if m > n:
        for i in range(m):
            for j in range(n, m):
                graph[i][j] = ciel[i]
        for i in range(m):
            for j in range(n):
                if jiro[j][0] == 'ATK':
                    if ciel[i] >= int(jiro[j][1]):
                        graph[i][j] = ciel[i] - int(jiro[j][1])
                    else:
                        graph[i][j] = -inf
                else:
                    if ciel[i] > int(jiro[j][1]):
                        graph[i][j] = 0
                    else:
                        graph[i][j] = -inf
        ans = max(ans, KM(m, graph).km())
    print(ans)